File: apps/services/event/src/loaders/services.loader.ts
/**
 * Builds the CQRS buses and registers the application services on them.
 */

import { createCommandBus, createQueryBus } from "@zudojs/cqrs";
import type { CommandBus, QueryBus } from "@zudojs/cqrs";
import type { Container } from "@zudojs/container";
import { LOGGER_TOKEN } from "../constants/index.js";
import { registerEventService } from "../services/index.js";

/** The buses the event service dispatches through. */
export interface EventBuses {
  readonly commandBus: CommandBus;
  readonly queryBus: QueryBus;
}

/**
 * Builds the command and query buses with every handler registered.
 *
 * @param container - The container the handlers' dependencies come from.
 * @returns The configured buses.
 */
export function loadServices(container: Container): EventBuses {
  const commandBus = createCommandBus();
  const queryBus = createQueryBus();

  registerEventService({ container, commandBus, queryBus });

  container.resolve(LOGGER_TOKEN).debug("CQRS handlers registered", {
    commands: commandBus.size(),
    queries: queryBus.size(),
  });

  return { commandBus, queryBus };
}

File: services/simulation/src/betng_simulation/loaders/services_loader.py
"""Builds the CQRS buses and registers the application services on them."""

from __future__ import annotations

from betng_service_kit import CommandBus, Container, QueryBus

from ..constants import LOGGER_TOKEN
from ..services import register_simulation_service


def load_services(container: Container) -> tuple[CommandBus, QueryBus]:
    """Build the buses with every handler registered.

    Args:
        container: The container the handlers' dependencies come from.

    Returns:
        The command bus and the query bus.
    """
    command_bus = CommandBus()
    query_bus = QueryBus()

    register_simulation_service(container, command_bus, query_bus)

    container.resolve(LOGGER_TOKEN).debug(
        "CQRS handlers registered",
        extra={"commands": command_bus.size(), "queries": query_bus.size()},
    )

    return command_bus, query_bus
